[chapter 3][용진] 3장 요구사항 구현

sirin0762 opened the pull request:
- 가장 최근의 커밋만 보면 될 것 같습니다
- BankStatmentProcessor의 경우 가장 마지막 메서드만 보시면 됩니다.
- 아직 테스트 코드는 작성하지 못했습니다 차후에 추가하여 반영하겠습니다.

sirin0762 committed:
[chapter 3].gitignore push

sirin0762 committed:
[chapter 3]chapter 2 내용

sirin0762 committed:
[chapter 3 요구사항 구현]
- Json 파일로 출력해주는 기능
- 특정 기간동안의 입출력 내역을 보여지는 기능
- 추가로 해야할 일 : 테스트 코드

minhee0327 commented on src/main/java/com/realsoftware/model/BankStatementProcessor.java:
```diff
@@ -0,0 +1,104 @@
+        return allSpending;
+    }
+
+    public List<Long> getTopTenSpending() {
+        List<Long> topTenSpending = new ArrayList<>();
+        List<BankTransaction> sortBySpendingDesc = sortBySpendingDesc();
+
+        int i = 0;
+        while (true) {
+            if (topTenSpending.size() == 10) {
+                break;
+            }
+            if (sortBySpendingDesc.get(i).getMoney() > 0) {
+                break;
+            }
+
+            topTenSpending.add(sortBySpendingDesc.get(i).getMoney());
+            i++;
+        }
+
+        return topTenSpending;
+    }
```
이부분을 한번 고민해보았는데요 !
리턴하는 타입을 List으로 금액정보(Long)만 보내주지 않고
List으로 거래내역정보(BankTransaction)를 모두 보내줘보는건 어떨까요 ?! 🤔

sirin0762 replied:
좋은 것 같습니다 ㅎㅎ 사용자가 금액만 궁금하진 않을 것 같네여!!
반영하겠습니다😁😁

minhee0327 commented on src/main/java/com/realsoftware/model/BankStatementProcessor.java:
```diff
@@ -0,0 +1,104 @@
+                .collect(Collectors.toList());
+    }
+
+    public MonthlyDetail getMonthlyDetail(int month) {
+        int monthlyDeposit = 0;
+        int monthlySpending = 0;
+        for (var bankTransaction : bankTransactions) {
+            LocalDate localDate = bankTransaction.getDate();
+            if (localDate.getMonthValue() == month) {
+                long money = bankTransaction.getMoney();
+                if (money > 0) {
+                    monthlyDeposit += 1;
+                } else {
+                    monthlySpending += 1;
+                }
+            }
+        }
+        return new MonthlyDetail(monthlyDeposit, monthlySpending);
+    }
```
int month의 방법도 좋은것 같아요!
저는 Month타입으로 받아서 구현을 했던 것 같은데 (저번에 얘기한 Month 타입의 이넘이 java에 내장되어있다구 신기했다구 했던?!)
이번에 과제에서 이넘으로 쓰는 방법을 많이 말씀해주셔서 한번 코멘트를 남겨봅니다아 ㅎㅎ

sirin0762 replied:
오오 맞네여 ㅎㅎ Month 로 바로 넣으면 더 좋을 것 같습니다!!

minhee0327 commented on test.json:
```diff
@@ -0,0 +1,29 @@
+[ {
+  "date" : "30-01-2017",
+  "money" : -100,
+  "location" : "Deliveroo"
+}, {
+  "date" : "30-01-2017",
+  "money" : -50,
+  "location" : "Tesco"
+}, {
+  "date" : "28-02-2017",
+  "money" : 6000,
+  "location" : "Salary"
+}, {
+  "date" : "28-02-2017",
+  "money" : 2000,
+  "location" : "Royalties"
+}, {
+  "date" : "30-03-2017",
+  "money" : -4000,
+  "location" : "Rent"
+}, {
+  "date" : "28-02-2017",
+  "money" : 3000,
+  "location" : "Tesco"
+}, {
+  "date" : "28-02-2017",
+  "money" : -30,
+  "location" : "Cinema"
+} ]
```
아직 json 타입의 파일을 처리하는 것을 구현하지 않아서 이부분이 조금 궁금했어요!
맨 밖에 []로 배열타입으로 지정하신 이유가 있으신지 궁금합니다 !

sirin0762 replied:
이게 json 파일로 쓰는 메서드를 쓰니까 저렇게 나오더라고여 테스트를 해서 이 파일이 json object가 되는지 확인해보겠습니다!

minhee0327 commented on src/main/java/com/realsoftware/model/BankStatementProcessor.java:
```diff
@@ -0,0 +1,104 @@
+        List<BankTransaction> result = new ArrayList<>();
+        for (var bankTransaction: bankTransactions){
+            LocalDate currentDate = bankTransaction.getDate();
+            if(!(currentDate.isBefore(startDate) || currentDate.isAfter(endDate))){
```
앗 이방법은 기발 한데요 ?ㅎㅎ 한 수 배워갑니다

sirin0762 replied:
감사합니다 ㅎㅎ😁😁

minhee0327 reviewed (commented):
전체적으로 변수명이나 메서드명이 잘 써있어서 읽기 편한 코드였어요!
늦게 리뷰한건 아니겠죵?ㅠㅠ

sirin0762 committed:
[chapter 3][FIX : 피드백 반영]
- Json 파일로 테스트 코드 추가
- 상위 10개의 리스트에 입출력 내역이 담기게 변경
- 추가로 해야할 일 : bankProcessor 테스트 코드
